{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "41fc721e",
   "metadata": {},
   "source": [
    "# Intersecting sgRNA sequences with VCF\n",
    "\n",
    "use bash scripts, i.e. `vcftools` and `bedtools`, to merge VCFs and intersect with the CDS-PAMs generated in step 1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b709f2c9",
   "metadata": {},
   "source": [
    "### Prepare VCF file\n",
    "\n",
    "1. Download Gnomad and dbSNP build 151:\n",
    "- Gnomad: https://gnomad.broadinstitute.org/downloads#v2-liftover-variants\n",
    "- dbSNP 151: https://ftp.ncbi.nlm.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/VCF/\n",
    "\n",
    "2. Strip `chr` in Gnomad VCF:\n",
    "```bash\n",
    "zcat gnomad.exomes.r2.1.1.sites.liftover_grch38.vcf.gz | awk '{gsub(/^chr/,\"\"); print}' | bgzip -c  > gnomad.exomes.r2.1.1.sites.liftover_grch38.NoCHR.vcf.gz ; tabix -p vcf gnomad.exomes.r2.1.1.sites.liftover_grch38.NoCHR.vcf.gz\n",
    "```\n",
    "\n",
    "3. Merge Gnomad and dbSNP\n",
    "```bash\n",
    "vcf-merge Gnomad_V2_Exomes_hg38_liftover/gnomad.exomes.r2.1.1.sites.liftover_grch38.NoCHR.vcf.gz hg38_All_20180418.vcf.gz  | bgzip -c > Gnomad_V2_Exomes_hg38_liftover-AND-hg38_dbSNP151_20180418.vcf.gz\n",
    "```\n",
    "**Warning: This could take a really long time (~2days).**\n",
    "\n",
    "\n",
    "4. Split by Chromosomes\n",
    "\n",
    "\n",
    "5. Intersection with CRISPR/Cas9 target sequences.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "130c0713",
   "metadata": {},
   "source": [
    "### To create .vcf files to intersect .bed files with (in 'human_SNP' NOT 'croton' folder)\n",
    "\n",
    "The chromosome level VCF files are generated by filtering out low MAF (<0.1%) variants:\n",
    "```bash\n",
    "for i in `seq 1 22`; do\n",
    "        echo $i\n",
    "        cat /mnt/ceph/users/vli/human_SNP/Gnomad_V2_Exomes_hg38_liftover2/chr$i.vcf | python /mnt/ceph/users/zzhang/croton/src/variant/filter_vcf.py > chr$i.vcf\n",
    "done\n",
    "\n",
    "cat /mnt/ceph/users/vli/human_SNP/Gnomad_V2_Exomes_hg38_liftover2/chrX.vcf | python /mnt/ceph/users/zzhang/croton/src/variant/filter_vcf.py > chrX.vcf\n",
    "```\n",
    "\n",
    "`filter_vcf.py` can be found in `backend` folder.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "de418cac",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "%%bash\n",
    "for i in `seq 1 22`; do \n",
    "    echo $i; \n",
    "    bedtools intersect \\\n",
    "    -a ./frontend/data/bed/CDSpams-byChrom/$i.bed \\\n",
    "    -b /mnt/home/zzhang/ceph/human_SNP/Gnomad_V2_Exomes_hg38_liftover/by_chrom/chr$i.vcf \\\n",
    "    -loj > ./frontend/data/bed/CDSpams-byChrom/$i.intersect.bed; \n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c470cebb",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "i=\"X\"\n",
    "echo $i; \n",
    "bedtools intersect \\\n",
    "-a ./frontend/data/bed/CDSpams-byChrom/$i.bed \\\n",
    "-b /mnt/home/zzhang/ceph/human_SNP/Gnomad_V2_Exomes_hg38_liftover/by_chrom/chr$i.vcf \\\n",
    "-loj > ./frontend/data/bed/CDSpams-byChrom/$i.intersect.bed; "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "af34f5cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext watermark\n",
    "%watermark -n -u -v -iv -w"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
